File: python/activities/llm.py
"""The PLAN step (slide 28, primitive 02): call the LLM as a Temporal Activity.

Two self-contained provider functions with the same signature — read the one
you care about and stop. `LLM_PROVIDER` picks one; the workflow never knows.

Retry philosophy (slide 31): the SDK's own retries are OFF (max_retries=0) so
Temporal owns ALL retries — every rate-limit or outage is visible as a retry
in workflow history. Requests the provider rejects outright (bad auth, bad
request) are marked non-retryable: retrying can't fix those.
"""

import json

import anthropic
import openai
from temporalio import activity
from temporalio.exceptions import ApplicationError

import config
from . import control
from models.types import ChatMessage, LLMRequest, LLMResponse, ToolCall
from prompts import TOOLS


@activity.defn
async def call_llm(req: LLMRequest) -> LLMResponse:
    # Demo kill-switch (the UI "API status" panel). While flipped to 'down',
    # raise a RETRYABLE error → Temporal retries with backoff until it's flipped
    # back, and the retries show in the history. This is the AI beat: LLM APIs
    # are flaky/rate-limited, and the loop rides it out with no extra code.
    if await control.llm_down():
        raise ApplicationError(
            "LLM provider is unavailable (simulated outage).", type="LLMProviderDown"
        )
    if config.LLM_PROVIDER == "openai":
        return await _call_openai(req)
    return await _call_anthropic(req)


async def _call_anthropic(req: LLMRequest) -> LLMResponse:
    client = anthropic.AsyncAnthropic(max_retries=0)

    system = next((m.content for m in req.messages if m.role == "system"), "")
    messages: list[dict] = []
    for m in req.messages:
        if m.role == "system":
            continue
        if m.role == "user":
            messages.append({"role": "user", "content": m.content})
        elif m.role == "assistant":
            blocks: list[dict] = []
            if m.content:
                blocks.append({"type": "text", "text": m.content})
            for c in m.tool_calls:
                blocks.append({"type": "tool_use", "id": c.id, "name": c.name, "input": c.args})
            messages.append({"role": "assistant", "content": blocks})
        else:
            block = {"type": "tool_result", "tool_use_id": m.tool_call_id, "content": m.content}
            last = messages[-1] if messages else None
            if (last and last["role"] == "user" and isinstance(last["content"], list)
                    and last["content"] and last["content"][-1].get("type") == "tool_result"):
                last["content"].append(block)
            else:
                messages.append({"role": "user", "content": [block]})

    try:
        resp = await client.messages.create(
            model=config.ANTHROPIC_MODEL,
            max_tokens=2048,
            system=system,
            tools=TOOLS,
            messages=messages,
        )
    except (anthropic.BadRequestError, anthropic.AuthenticationError,
            anthropic.PermissionDeniedError, anthropic.NotFoundError) as e:
        raise ApplicationError(f"LLM request rejected: {e}", type="LLMFatalError", non_retryable=True) from e

    text = "".join(b.text for b in resp.content if b.type == "text")
    calls = [ToolCall(id=b.id, name=b.name, args=b.input)
             for b in resp.content if b.type == "tool_use"]
    return LLMResponse(message=ChatMessage(role="assistant", content=text, tool_calls=calls))


async def _call_openai(req: LLMRequest) -> LLMResponse:
    client = openai.AsyncOpenAI(max_retries=0)

    messages: list[dict] = []
    for m in req.messages:
        if m.role == "tool":
            messages.append({"role": "tool", "tool_call_id": m.tool_call_id,
                             "content": m.content})
        elif m.role == "assistant" and m.tool_calls:
            messages.append({
                "role": "assistant",
                "content": m.content or None,
                "tool_calls": [{"id": c.id, "type": "function",
                                "function": {"name": c.name, "arguments": json.dumps(c.args)}}
                               for c in m.tool_calls],
            })
        else:
            messages.append({"role": m.role, "content": m.content})

    tools = [{"type": "function",
              "function": {"name": t["name"], "description": t["description"],
                           "parameters": t["input_schema"]}}
             for t in TOOLS]

    try:
        resp = await client.chat.completions.create(
            model=config.OPENAI_MODEL, messages=messages, tools=tools
        )
    except (openai.BadRequestError, openai.AuthenticationError,
            openai.PermissionDeniedError, openai.NotFoundError) as e:
        raise ApplicationError(f"LLM request rejected: {e}", type="LLMFatalError", non_retryable=True) from e

    choice = resp.choices[0].message
    calls = [ToolCall(id=c.id, name=c.function.name, args=json.loads(c.function.arguments))
             for c in (choice.tool_calls or [])]
    return LLMResponse(message=ChatMessage(role="assistant", content=choice.content or "",
                                           tool_calls=calls))

File: typescript/src/activities/llm.ts
// The PLAN step: call the LLM as a Temporal Activity. Anthropic-only (the TS
// worker doesn't carry the OpenAI path). Ported from python/activities/llm.py.
//
// Retry philosophy: the SDK's own retries are OFF (maxRetries: 0) so Temporal
// owns ALL retries — every rate-limit or outage is visible as a retry in
// workflow history. Requests the provider rejects outright (bad auth/request)
// are marked non-retryable: retrying can't fix those.

import Anthropic from '@anthropic-ai/sdk';
import { ApplicationFailure } from '@temporalio/common';

import * as config from '../config';
import * as control from './control';
import { TOOLS } from '../prompts';
import type { LLMRequest, LLMResponse, ToolCall } from '../types';

// HTTP statuses the provider rejects outright — retrying can't fix these.
const FATAL_STATUS = new Set([400, 401, 403, 404]);

export async function callLlm(req: LLMRequest): Promise<LLMResponse> {
  // Demo kill-switch (the UI "API status" panel). While flipped to 'down', raise
  // a RETRYABLE error → Temporal retries with backoff until it's flipped back,
  // and the retries show in the history.
  if (await control.llmDown()) {
    throw ApplicationFailure.create({
      message: 'LLM provider is unavailable (simulated outage).',
      type: 'LLMProviderDown',
    });
  }
  return callAnthropic(req);
}

async function callAnthropic(req: LLMRequest): Promise<LLMResponse> {
  const client = new Anthropic({ maxRetries: 0 });

  const system = req.messages.find((m) => m.role === 'system')?.content ?? '';
  const messages: Anthropic.MessageParam[] = [];
  for (const m of req.messages) {
    if (m.role === 'system') continue;
    if (m.role === 'user') {
      messages.push({ role: 'user', content: m.content });
    } else if (m.role === 'assistant') {
      const blocks: Anthropic.ContentBlockParam[] = [];
      if (m.content) blocks.push({ type: 'text', text: m.content });
      for (const c of m.tool_calls ?? []) {
        blocks.push({ type: 'tool_use', id: c.id, name: c.name, input: c.args });
      }
      messages.push({ role: 'assistant', content: blocks });
    } else {
      // role === 'tool' → an Anthropic tool_result, batched into the trailing
      // user turn if the previous message already ended with one.
      const block: Anthropic.ToolResultBlockParam = {
        type: 'tool_result',
        tool_use_id: m.tool_call_id!,
        content: m.content,
      };
      const last = messages[messages.length - 1];
      if (
        last &&
        last.role === 'user' &&
        Array.isArray(last.content) &&
        last.content.length &&
        (last.content[last.content.length - 1] as Anthropic.ContentBlockParam).type === 'tool_result'
      ) {
        (last.content as Anthropic.ContentBlockParam[]).push(block);
      } else {
        messages.push({ role: 'user', content: [block] });
      }
    }
  }

  let resp: Anthropic.Message;
  try {
    resp = await client.messages.create({
      model: config.ANTHROPIC_MODEL,
      max_tokens: 2048,
      system,
      tools: TOOLS as unknown as Anthropic.Tool[],
      messages,
    });
  } catch (e) {
    if (e instanceof Anthropic.APIError && typeof e.status === 'number' && FATAL_STATUS.has(e.status)) {
      throw ApplicationFailure.create({
        message: `LLM request rejected: ${e.message}`,
        type: 'LLMFatalError',
        nonRetryable: true,
      });
    }
    throw e;
  }

  let text = '';
  const calls: ToolCall[] = [];
  for (const b of resp.content) {
    if (b.type === 'text') text += b.text;
    else if (b.type === 'tool_use') {
      calls.push({ id: b.id, name: b.name, args: (b.input ?? {}) as Record<string, unknown> });
    }
  }
  return { message: { role: 'assistant', content: text, tool_calls: calls } };
}
